File: leetcode/513.找树左下角的值.go
package main

import "fmt"

/*
 * @lc app=leetcode.cn id=513 lang=golang
 *
 * [513] 找树左下角的值
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start

func findBottomLeftValue(root *TreeNode) int {
	nodeList := make([]*TreeNode, 0)
	nodeList = append(nodeList, root)
	var res *TreeNode
	for len(nodeList) > 0 {
		res = nodeList[0]
		newNodeList := make([]*TreeNode, 0)
		for _, node := range nodeList {
			if node.Left != nil {
				newNodeList = append(newNodeList, node.Left)
			}
			if node.Right != nil {
				newNodeList = append(newNodeList, node.Right)
			}
		}
		nodeList = newNodeList
	}
	return res.Val
}

// @lc code=end
func main() {
	fmt.Println(findBottomLeftValue(&TreeNode{2, &TreeNode{1, nil, nil}, &TreeNode{3, nil, nil}}))
}

File: leetcode/513.找树左下角的值.cpp
#include <iostream>
#include <list>
using namespace std;
/*
 * @lc app=leetcode.cn id=513 lang=cpp
 *
 * [513] 找树左下角的值
 */
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
class Solution
{
public:
    int findBottomLeftValue(TreeNode *root)
    {
        list<TreeNode *> nodeList;
        nodeList.push_back(root);
        TreeNode *res;
        while (!nodeList.empty())
        {
            res = nodeList.front();
            int n = nodeList.size();
            for (int i = 0; i < n; i++)
            {
                TreeNode *node = nodeList.front();
                nodeList.pop_front();
                if (node->left != nullptr)
                {
                    nodeList.push_back(node->left);
                }
                if (node->right != nullptr)
                {
                    nodeList.push_back(node->right);
                }
            }
        }
        return res->val;
    }
};
// @lc code=end
int main()
{
    TreeNode node1 = TreeNode(1);
    TreeNode node3 = TreeNode(3);
    TreeNode node2 = TreeNode(2, &node1, &node3);
    cout << Solution().findBottomLeftValue(&node2) << endl;
    return 0;
}
